src/Transports.AspNetCore.SystemTextJson/GraphQLRequestDeserializer.cs

```diff
@@ -0,0 +1,117 @@
+using GraphQL.Server.Transports.AspNetCore.Common;
+using GraphQL.SystemTextJson;
+using Microsoft.AspNetCore.Http;
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using GraphQLRequestBase = GraphQL.Server.Common.GraphQLRequest;
+
+namespace GraphQL.Server.Transports.AspNetCore.SystemTextJson
+{
+    /// <summary>
+    /// Implementation of an <see cref="IGraphQLRequestDeserializer"/> that uses System.Text.Json;
+    /// reading the request body asynchronously and optimally.
+    /// </summary>
+    /// <remarks>
+    /// With thanks to David Fowler (@davidfowl) for his help getting this right.
+    /// </remarks>
+    public class GraphQLRequestDeserializer : IGraphQLRequestDeserializer
+    {
+        private readonly JsonSerializerOptions _serializerOptions = new JsonSerializerOptions();
+
+        public GraphQLRequestDeserializer(Action<JsonSerializerOptions> configure)
+        {
+            _serializerOptions.Converters.Add(new ObjectDictionaryConverter());
+
+            configure?.Invoke(_serializerOptions);
+        }
+
+        public async Task<GraphQLRequestDeserializationResult> DeserializeFromJsonBodyAsync(HttpRequest httpRequest, CancellationToken cancellationToken = default)
+        {
+            JsonTokenType jsonTokenType;
+            try
+            {
+                jsonTokenType = await PeekJsonTokenTypeAsync(httpRequest.BodyReader, cancellationToken).ConfigureAwait(false);
+            }
+            catch (JsonException)
+            {
+                // Invalid request content, assign None so it falls through to WasSuccessful = false
+                jsonTokenType = JsonTokenType.None;
+            }
+
+            var result = new GraphQLRequestDeserializationResult() { IsSuccessful = true };
+            switch (jsonTokenType)
+            {
+                case JsonTokenType.StartObject:
+                    result.Single = await JsonSerializer.DeserializeAsync<GraphQLRequest>(httpRequest.BodyReader.AsStream(), _serializerOptions, cancellationToken)
+                        .ConfigureAwait(false);
+                    return result;
+                case JsonTokenType.StartArray:
+                    result.Batch = await JsonSerializer.DeserializeAsync<GraphQLRequest[]>(httpRequest.BodyReader.AsStream(), _serializerOptions, cancellationToken)
+                        .ConfigureAwait(false);
+                    return result;
+                default:
+                    result.IsSuccessful = false;
+                    return result;
+            }
+        }
+
+        private static async ValueTask<JsonTokenType> PeekJsonTokenTypeAsync(PipeReader reader, CancellationToken cancellationToken = default)
```

Comment:
What should I do with this cancellation token for this method? Check for cancelled in the `while` loop and if `true`, do something? What exactly? Return `JsonTokenType.None` since it doesn't matter?

Reply:
`token.ThrowIfCancellationRequested();` should be enough.

Reply:
Added.